x-pack/plugins/encrypted_saved_objects/server/crypto/encryption_key_rotation_service.ts

```diff
@@ -0,0 +1,252 @@
+      excludedWrappers: ['encryptedSavedObjects'],
+    });
+    const updateClient = savedObjects.getScopedClient(request, {
+      includedHiddenTypes: registeredHiddenSavedObjectTypes,
+    });
+
+    // Keeps track of object IDs that have been processed already.
+    const processedObjectIDs = new Set<string>();
+
+    // Until we get scroll/search_after support in Saved Objects client we have to retrieve as much objects as allowed
+    // by the `batchSize` parameter. Instead of using paging functionality (size/from  or page/perPage parameters) that
+    // has certain performance issues and is also limited by the maximum result window setting on .kibana index
+    // (10,000 by default) we always fetch the first page of the results sorted by the `updated_at` field. This way we
+    // can prioritize "old" objects that have a higher chance to have been encrypted with the old encryption keys, since
+    // all newly created or updated objects are always encrypted with the current primary key. Re-encryption of the
+    // "old" objects with the primary key implicitly bumps up their `updated_at` field so that these objects won't be
+    // included into the first page of the results during next iteration. Additionally we track IDs of all processed
+    // objects so that eventually we can detect that first page consists of only objects encrypted with the current
+    // primary key and stop iterating.
+    //
+    // LIMITATION: if we have a lot of "old" objects encrypted with the _unknown_ encryption key it may either
+    // significantly slow down rotation or prevent it from happening completely since such objects will be included into
+    // every batch we fetch and if their number is equal to or greater than `batchSize` we won't be able to process any
+    // object. Another and more complex case when we can be hit by this limitation is when users have multiple Kibana
+    // instances configured with different primary encryption keys, these time even "new" objects may require rotation,
+    // but they may be included into 2+ page of the results. We can potentially detect such cases and issue a warning,
+    // but it's not an easy task: if we detect a case when none of the objects from the very first batch cannot be
+    // decrypted with the decryption only keys we'll need to check how many of them can be decrypted at all using all
+    // available keys including the current primary one.
+    //
+    // Also theoretically if `batchSize` is less than `index.max_result_window` we could try to rely on the paging
+    // functionality and switch to the second page, but the issue here is that objects can be deleted in the meantime
+    // so that unprocessed objects may get into the first page and we'll miss them. We can of course oscillate between
+    // the first and the second pages or do multiple rotation passes, but it'd complicate code significantly.
+    let batch = 0;
+    while (true) {
```

Comment:
> I worry about having this loop terminate based solely on user data. What do you think about introducing a safeguard to only allow this loop to iterate a certain number of times before terminating?

I was thinking about the same initially, but I could not come up with the real case when this can turn into infinite loop:

* if `find` throws - we'll throw
* if `find` doesn't return any object, we'll `break` at line 167 (`if (savedObjectsToEncrypt.length === 0) { break; }`)
* if `find` returns only objects we already processed then `getSavedObjectsToReEncrypt` will return empty array and we'll `break` at line 167 (`if (savedObjectsToEncrypt.length === 0) { break; }`)

The only case when we can stuck in the loop only is if we'll be getting new objects all the time. Are you concerned about this case or you have another one in mind? I'm fine with adding this hard limitation though, just in case.

Reply:
>  then they will be racing against each other and will continually find objects that need updating.

That shouldn't be the problem though, since IDs don't change here and we'll skip them (since we remember IDs we already processed).

Reply:
> That shouldn't be the problem though, since IDs don't change here and we'll skip them (since we remember IDs we already processed).

Ah of course 🙈

Reply:
Good, I'll add this check then. We have a theoretical use case (e.g. objects are constantly created), so it makes sense to guard against it.